### PR TITLE
[NPU] fix fp16 (PART I)

### PR DESCRIPTION
**Pull request by windstamp**
### PR types

Bug fixes
### PR changes

OPs
### Describe

fix fp16 for npu.

### CODE REVIEW

File: paddle/fluid/operators/batch_norm_op_npu.cc [hunk 1]
```diff
@@ -76,10 +76,10 @@ class NPUBatchNormOpKernel : public framework::OpKernel<T> {
...
       auto *saved_mean = ctx.Output<Tensor>("SavedMean");
       auto *saved_variance = ctx.Output<Tensor>("SavedVariance");
-      mean_out->mutable_data<T>(ctx.GetPlace());
-      variance_out->mutable_data<T>(ctx.GetPlace());
-      saved_mean->mutable_data<T>(ctx.GetPlace());
-      saved_variance->mutable_data<T>(ctx.GetPlace());
+      mean_out->mutable_data<float>(ctx.GetPlace());
+      variance_out->mutable_data<float>(ctx.GetPlace());
+      saved_mean->mutable_data<float>(ctx.GetPlace());
+      saved_variance->mutable_data<float>(ctx.GetPlace());
```
Comment (qili93) [inline paddle/fluid/operators/batch_norm_op_npu.cc:82]: 在FP32/FP16下把这里weight值都改成float类型，需要确认2种情况
1. 跑AMP O1/O2下，能否正常保存和读取的checkpoint（即weight值存成FP32会不会读错）
2. 跑Inference量化模型的时候，预测结果是否正确（即weight值转成FP32会不会错误）

Comment (windstamp) [reply paddle/fluid/operators/batch_norm_op_npu.cc:82]: AMP O2 目前还有问题，暂时跑不通。

AMP O1 测试 PaddleClas ResNet50 在小数据集 flowers102 断点重新训练，在 Epoch 17 主动中断再重新训练，测试结果表明，断点重新训练是正常的，最终的精度和非断点重新训练是一致的。
图片

图片

图片

[CUDA 实现](https://github.com/PaddlePaddle/Paddle/blob/v2.2.2/paddle/fluid/operators/batch_norm_op.cu#L328) 也是把这些 weight 值设置成 FP32 的（对于 FP16 的输入参数）。

同时，根据 [BatchNormOp::GetExpectedKernelType](https://github.com/PaddlePaddle/Paddle/blob/v2.2.2/paddle/fluid/operators/batch_norm_op.cc#L153)，对于 FP16 的输入参数，这些 weight 输入时也必须是 FP32。

**Review by qili93 — APPROVED** [review]
LGTM
